python/pyarrow/parquet.py

```diff
@@ -152,6 +153,8 @@ def _visit_piece(i, key, rest):
         for i, path in enumerate(paths):
             _visit_piece(i, path[0], path[1:])
 
+        _visit_piece = None  # Remove GC cycle.
```

Comment:
This looks a bit convoluted to me. Just make `_visit_piece` iterative rather than recursive and the cycle should disappear.